system/jlib/jqueue.tpp

```diff
@@ -547,4 +547,98 @@ public:
                                     while (x--) 
 
 
+template <class BASE>
+class RingBufferOf
+{
+    typedef RingBufferOf<BASE> SELF;
+    BASE **ptrs;
+    unsigned headp;
+    unsigned tailp;
+    unsigned max;
+    unsigned num;
+
+    virtual void releaseBuffer(BASE *e){};
+    virtual void releaseBufferAll()
+    {
+        while (num)
+        {
+            releaseBuffer(ptrs[headp]);
+            headp++;
+            num--;
+            if (headp==max) headp=0;
+        }
+    };
+
+public:
+    inline RingBufferOf() : ptrs(NULL), headp(0), tailp(0), max(0), num(0){};
+
+    virtual ~RingBufferOf() { clear(); }
+
+    inline void clear()
+    {
+        releaseBufferAll();
+        free(ptrs);
+        ptrs=NULL;
+        headp=0;
+        tailp=0;
+        max=0;
+        num=0;
+    }
+    inline void ensure(unsigned bufSize)
+    {
+        clear();
```

Comment:
Not sure it is really necessary to require that changing the size of the ring buffer also empties it, and I can foresee situations when you might prefer that it did not.